File: src/main/java/tech/bobliu/assignment07/controller/UserServlet.java
package tech.bobliu.assignment07.controller;

import com.google.gson.Gson;
import jakarta.servlet.annotation.WebServlet;
import jakarta.servlet.http.HttpServlet;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;
import tech.bobliu.assignment07.model.User;
import tech.bobliu.assignment07.service.UserService;

import java.io.IOException;
import java.io.Serializable;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

@WebServlet(name = "userServlet", value = {"/user/*"})
public class UserServlet extends HttpServlet {
    private final Gson gson = new Gson();

    @Override
    protected void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
        String path = request.getPathInfo();
        if ("/online".equals(path)) {
            response.setContentType("application/json;charset=UTF-8");
            User user = (User) request.getSession().getAttribute("user");
            List<Map<String, ? extends Serializable>> onlineUsers = UserService.getOnlineUsers().stream().map(u -> Map.of(
                    "id", u.getId(),
                    "username", u.getUsername(),
                    "isSelf", u.getId() == user.getId()
            )).collect(Collectors.toList());

            response.getWriter().write(gson.toJson(Map.of("status", "success", "users", onlineUsers)));
        }
    }
}

File: src/main/kotlin/tech/bobliu/assignment07/controller/UserServlet.kt
package tech.bobliu.assignment07.controller

import com.google.gson.Gson
import jakarta.servlet.annotation.WebServlet
import jakarta.servlet.http.HttpServlet
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import tech.bobliu.assignment07.model.User
import tech.bobliu.assignment07.service.UserService

@WebServlet(name = "userServlet", value = ["/user/*"])
class UserServlet : HttpServlet() {
    private val gson = Gson()

    override fun doGet(request: HttpServletRequest, response: HttpServletResponse) {
        when (request.pathInfo) {
            "/online" -> {
                response.contentType = "application/json;charset=UTF-8"
                val user = request.session.getAttribute("user") as User
                val onlineUsers = UserService.getOnlineUsers().map {
                    mapOf(
                        "id" to it.id,
                        "username" to it.username,
                        "isSelf" to (it.id == user.id)
                    )
                }

                response.writer.write(gson.toJson(mapOf("status" to "success", "users" to onlineUsers)))
            }
        }
    }
}